tendermint-testgen/src/helpers.rs

```diff
@@ -0,0 +1,64 @@
+//! Helper functions
+
+use serde::de::DeserializeOwned;
+use signatory::signature::Verifier;
+use signatory_dalek::Ed25519Verifier;
+use simple_error::*;
+use std::io::{self, Read};
+use tendermint::{amino_types, signature::Signature, vote};
+
+/// A macro that generates a complete setter method from a one-liner with necessary information
+#[macro_export]
+macro_rules! set_option {
+    ($name:ident, $t:ty) => {
+        pub fn $name(mut self, $name: $t) -> Self {
+            self.$name = Some($name.clone());
+            self
+        }
+    };
+    ($name:ident, $t:ty, $val:expr) => {
+        pub fn $name(mut self, $name: $t) -> Self {
+            self.$name = $val;
+            self
+        }
+    };
+}
+
+/// Tries to parse a string as the given type; otherwise returns the input wrapped in SimpleError
+pub fn parse_as<T: DeserializeOwned>(input: &str) -> Result<T, SimpleError> {
+    match serde_json::from_str(input) {
+        Ok(res) => Ok(res),
+        Err(_) => Err(SimpleError::new(input)),
+    }
+}
+
+pub fn read_stdin() -> Result<String, SimpleError> {
+    let mut buffer = String::new();
+    try_with!(io::stdin().read_to_string(&mut buffer), "");
+    Ok(buffer)
+}
+
+/// Tries to parse STDIN as the given type; otherwise returns the input wrapped in SimpleError
+pub fn parse_stdin_as<T: DeserializeOwned>() -> Result<T, SimpleError> {
+    let mut buffer = String::new();
+    match io::stdin().read_to_string(&mut buffer) {
+        Err(_) => Err(SimpleError::new("")),
```

Comment:
It would be nice to keep the underlying error around, even if just as a string, to ease debugging a bit perhaps.

Reply:
Also `read_stdin` and `parse_stdin_as` could perhaps be rewritten in the same style, ie. with `try_with!` or with a `match`.

Reply:
`parse_stdin_as` was not used anymore actually; so I removed it; and `read_stdin` I've rewritten using `match`